File: Prime age calc/pythoncode/primeAgeCalculator.py
import datetime

# Function to check if a number is prime
def is_prime(n):
    if n <= 1: 
        return False  # Numbers less than 2 are not prime numbers
    if n <= 3: 
        return True  # 2 and 3 are prime numbers
    if n % 2 == 0 or n % 3 == 0: 
        return False  # Numbers divisible by 2 or 3 are not prime
    i = 5
    while i * i <= n:  # We only need to check divisibility up to the square root of n
        if n % i == 0 or n % (i + 2) == 0: 
            return False  # If n is divisible by any number, it's not prime
        i += 6
    return True  # If n is not divisible by any number, it's prime

# Function to calculate the next year when the person's age will be a prime number
def next_prime_age_year(birth_year):
    current_year = datetime.date.today().year  # Get the current year
    age = current_year - birth_year  # Calculate the person's current age
    while True:  # Keep looping until we find a prime age
        age += 1  # Increment the age
        if is_prime(age):  # Check if the incremented age is prime
            return birth_year + age  # If it is, return the year when the person will be of that age

if __name__ == "__main__":
    while True:  # Start an infinite loop
        birth_year = input("Please enter your birth year or 'q' to quit: ")  # Ask for input
        if birth_year.lower() == 'q':  # If the user entered 'q' or 'Q', break the loop
            break
        birth_year = int(birth_year)  # Convert the input to an integer
        print("The next year when your age will be a prime number is: ", next_prime_age_year(birth_year))  # Calculate and print the result

File: Prime age calc/javacode/PrimeAgeCalculator.java
// Import necessary libraries
import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.time.Year;

// Class declaration
public class PrimeAgeCalculator {

    // Declare GUI components
    private JFrame frame;
    private JTextField textFieldBirthYear;
    private JTextArea textAreaResult;

    // Class constructor
    public PrimeAgeCalculator() {
        // Initialize JFrame with title
        frame = new JFrame("Prime Age Calculator");
        // Set default close operation for JFrame
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        // Set size of the JFrame
        frame.setSize(300, 200);

        // Initialize JTextField and JTextArea
        textFieldBirthYear = new JTextField();
        textAreaResult = new JTextArea();

        // Initialize JButton and its ActionListener
        JButton buttonCalculate = new JButton("Calculate");
        buttonCalculate.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                // Check if textFieldBirthYear is not empty
                if (!textFieldBirthYear.getText().isEmpty()) {
                    // Convert the birthYear from string to int
                    int birthYear = Integer.parseInt(textFieldBirthYear.getText());
                    // Calculate the next year of prime age
                    int nextPrimeYear = nextPrimeAgeYear(birthYear);
                    // Display the next year of prime age in textAreaResult
                    textAreaResult.setText("Next prime age year for you will be: " + nextPrimeYear);
                } else {
                    // Display error message in textAreaResult if textFieldBirthYear is empty
                    textAreaResult.setText("Please enter a valid birth year.");
                }
            }
        });

        // Add JTextField, JButton and JTextArea to JFrame
        frame.getContentPane().add(textFieldBirthYear, "North");
        frame.getContentPane().add(buttonCalculate, "Center");
        frame.getContentPane().add(textAreaResult, "South");
    }

    // Method to display the JFrame
    public void show() {
        frame.setVisible(true);
    }

    // Method to calculate the next year of prime age
    public int nextPrimeAgeYear(int birthYear) {
        // Get the current year and calculate age
        int currentYear = Year.now().getValue();
        int age = currentYear - birthYear;

        // Keep incrementing the age until it is a prime number
        while (true) {
            age++;
            if (isPrime(age)) {
                // Return the year when age will be a prime number
                return birthYear + age;
            }
        }
    }

    // Method to check if a number is prime
    public boolean isPrime(int n) {
        // Check if number is less than or equal to 1
        if (n <= 1) {
            return false;
        }
        // Check if number is less than or equal to 3
        if (n <= 3) {
            return true;
        }
        // Check if number is divisible by 2 or 3
        if (n % 2 == 0 || n % 3 == 0) {
            return false;
        }
        // Check if number is divisible by any number from 5 to sqrt(n)
        for (int i = 5; i * i <= n; i = i + 6) {
            if (n % i == 0 || n % (i + 2) == 0) {
                return false;
            }
        }
        // Return true if number is prime
        return true;
    }

    // Main method to create and display the PrimeAgeCalculator GUI
    public static void main(String[] args) {
        PrimeAgeCalculator calculator = new PrimeAgeCalculator();
        calculator.show();
    }
}
